### PR TITLE
add package NumericalSemigroups

### PR DESCRIPTION
**Pull request by eisenbud**
New package created by Frank Schreyer and David Eisenbud with basic commands for analyzing semigroups using
the Kunz cones
and studying when they are Weierstrass semigroups.

### CODE REVIEW

**Comment by eisenbud** [comment]
Also updated "PencilsOfQuadrics"

File: M2/Macaulay2/packages/PencilsOfQuadrics.m2 [hunk 1]
```diff
@@ -9,25 +9,26 @@ viewHelp "PencilsOfQuadrics"
...
                        HomePage => ""},
              	         {Name => "David Eisenbud", 
                        Email => "de@msri.org", 
-                       HomePage => "https://www.msri.org/~de"},
+                       HomePage => "www.msri.org/~de"},
 		       {Name => "Yeongrak Kim",
-			   Email => "kim@math.uni-sb.de",
-			   HomePage => "https://sites.google.com/view/yeongrak"}
+    	    	    	   Email => "yeongrak.kim@pusan.ac.kr",
+			   HomePage => "sites.google.com/view/yeongrak"}
```
Comment (d-torrance) [inline M2/Macaulay2/packages/PencilsOfQuadrics.m2:26]: Removing the "https://" from these two `HomePage` fields is causing a failing test:

```
cd /home/runner/work/M2/M2/M2/BUILD/build/usr-dist/common/share/doc/Macaulay2 && find . -name \*.html | xargs -r ls -rt | LD_LIBRARY_PATH="/home/runner/work/M2/M2/M2/BUILD/build/usr-host/lib:$LD_LIBRARY_PATH" xargs "/home/runner/work/M2/M2/M2/BUILD/build/Macaulay2/html-check-links"/html-check-links --no-absolute-links
./PencilsOfQuadrics/html/index.html:124:9: error: broken link: ../../../../../../www.msri.org/~de
./PencilsOfQuadrics/html/index.html:126:9: error: broken link: ../../../../../../sites.google.com/view/yeongrak
make[2]: *** [Makefile:62: run-html-check-links] Error 123
```

**Comment by eisenbud** [comment]
many tests in Pencil... used random numbers and thus failed randomly. I've inserted setRandomSeed commands to fix this.

**Comment by d-torrance** [comment]
The random seed is always set to 0 by default for tests and examples, so I don't think the most recent commit is necessary.

The failing build from earlier was unrelated to this PR (#1579).

**Comment by eisenbud** [comment]
Thanks, Doug -- This is my first experience(s) doing a pull request by myself. I had assumed that the failing sumsOfSquares call indicated in the "Details" was something indirectly called from my package. Should I have just ignored the failure?
-David

**Comment by d-torrance** [comment]
Often build failures are related to the PR (like the `Homepage` one earlier), so I wouldn't say you should ignore them in general.

If you're not sure, then you can search through the existing issues to see the failure is a known problem like this one, or just ask!

**Comment by mahrud** [comment]
Usually if only a single build fails it could be one of these unrelated but persistent failures that show up randomly, but if two fail probably one of the tests failed and four fail it's probably a failing example.

Perhaps at some point we can have GitHub annotations #3533 to easily tell us what failed without having to dig through the build log.

**Comment by d-torrance** [comment]
@eisenbud - Did you want to keep the `setRandomSeed` commit?  If not, you can run the following:

```
git reset --hard HEAD~
git push --force <remote> development
```

`HEAD~` refers to the commit immediately preceding the current one.  Replace `<remote>` with whatever name you've given your fork.

**Comment by mikestillman** [comment]
@d-torrance @eisenbud The one failing test (I think) is the SumOfSquares failure, which should not be related to this PR.
Our Complexes PR (#3526) will make changes to PencilOfQuadrics too.  This PR should be done before #3526, that way, I can deal with the resulting conflicts.

**Comment by d-torrance** [comment]
> The one failing test (I think) is the SumOfSquares failure, which should not be related to this PR.

That's correct -- see #1579.  @mahrud has a fix in #3519

I'm not sure if this should be merged now or not.  The final commit (3e672a632599aa32f9e9da99616207162aab119b) may not be necessary.

**Comment by mahrud** [comment]
David, do you mind if I force push a changes to your branch to remove the last commit and fix the failing build?

File: M2/Macaulay2/packages/PencilsOfQuadrics.m2 [hunk 2]
```diff
@@ -700,11 +728,13 @@ randomLineBundle(ZZ, RingElement) := (d,f) -> (
...
---load "PencilsOfQuadrics.m2"
-///
-///--test of preRandomLineBundle and randomLineBundle-- but there were no "asserts"
+
+
+TEST///
+--test of preRandomLineBundle and randomLineBundle
+restart
+load "PencilsOfQuadrics.m2"
+>>>>>>> 82c56807936dda91b86f9ccae7189b001e15ff19
```
Comment (mahrud) [inline M2/Macaulay2/packages/PencilsOfQuadrics.m2:737]: This line is going to cause errors (and also the "restart" two lines above will prevent this test from running)

**Comment by d-torrance** [comment]
The build failure is unrelated (#3239).  Merging!

**Pull request merged**